introduce FEInterfaceValues

tjhei opened the pull request:
part 1 of many, see #8541 for details.

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+ * FEInterfaceValues is a data structure to access and assemble Finite Element
+ * data on interfaces between two cells of a mesh.
+ *
+ * Basically, it provides a way to access average and jump terms used in
```
```suggestion
 * It provides a way to access average and jump terms used in
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+ * The class is made to be used inside MeshWorker::mesh_loop. It is intended
+ * to be a low level replacement for MeshWorker and LocalIntegrators and a
+ * higher level abstraction compared to assembling face terms manually.
+ */
```
```suggestion
 * 
 * @author Timo Heister, 2019.
 */
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+  FEInterfaceValues(const Mapping<dim, spacedim> &      mapping,
+                    const FiniteElement<dim, spacedim> &fe,
+                    const Quadrature<dim - 1> &         quadrature,
+                    const UpdateFlags                   update_flags);
```
Consistently to what we do in other `FEValues` classes, can you add a constructor that does not take a mapping, and just calls `StaticMappingQ1<dim, spacedim>::mapping`?

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+   * Re-initialize this object to be used on a new interface given by two faces
+   * of two neighboring cells.
+   *
+   * The arguments including their order is identical to the @p face_worker arguments
```
```suggestion
   * The arguments (including their order) is identical to the @p face_worker arguments
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+  get_quadrature() const;
+
+  /**
+   * Return a reference to the quadrature object in use.
```
```suggestion
   * Return a reference to the quadrature points.
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+  const UpdateFlags
+  get_update_flags() const;
+
+
```
```suggestion
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+                    const Quadrature<dim - 1> &         quadrature,
+                    const UpdateFlags                   update_flags);
+
+
```
```suggestion
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+  n_interface_dofs() const;
+
+  /**
+   * Return if the current interface, which is set by reinit(), is an internal
```
```suggestion
   * Return if the current interface is an internal
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+  /**
+   * Return the FEFaceValue object of the cell or the neighboring cell
+   *
+   * If @p cell_or_neighbor is 0, return the cell's, if it is 1, return the
```
```suggestion
   * If @p cell_or_neighbor is 0, return the cell FEFaceValues object, if it is 1, return the
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+   * Return the FEFaceValue object of the cell or the neighboring cell
+   *
+   * If @p cell_or_neighbor is 0, return the cell's, if it is 1, return the
+   * neighboring cell's FEFaceValue object.
```
```suggestion
   * neighboring cell FEFaceValue object.
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+  get_fe_values(const unsigned int cell_or_neighbor) const;
+
+  /**
+   * Return the FEFaceValue object of the current cell
```
```suggestion
   * Return the FEFaceValues object of the current cell
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+  get_fe_values() const;
+
+  /**
+   * Return the FEFaceValue object of the current neighboring cell
```
```suggestion
   * Return the FEFaceValues object of the current neighboring cell
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+  normal(const unsigned int q_point_index) const;
+
+  /**
+   * Return the current (if @p current_cell is true) or neighboring (if @p
```
```suggestion
   * Return the current value (if @p current_cell is true) or the neighboring value (if @p
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+
+  /**
+   * Return the current (if @p current_cell is true) or neighboring (if @p
+   * current_cell is false) value on the interface for the shape function @p
```
```suggestion
   * current_cell is false) on the interface for the shape function @p
```

luca-heltai commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,608 @@
+         const unsigned int component = 0) const;
+
+  /**
+   * Return the jump $[u]=u_1 - u_2$ on the interface for the shape function
```
I don't think $1$ and $2$ have any meaning in this context. Maybe use `cell` and `neighbor`?
```suggestion
   * Return the jump $[u]=u_{\text{cell}} - u_{\text{neighbor}}$ on the interface for the shape function
```

luca-heltai commented on tests/feinterface/step-12.cc:
```diff
@@ -0,0 +1,580 @@
+        }
+    };
+
+    /*
```
If you don't need this, can you remove it, or can you expand with a comment what this does differently from the other boundary worker?

tjhei replied:
I tried to use FEInterfaceValues instead of FEValues for the boundary, but it doesn't make a lot of sense. Code is removed.

luca-heltai commented on tests/feinterface/step-12.cc:
```diff
@@ -0,0 +1,580 @@
+                           CopyData &          copy_data) {
+      FEInterfaceValues<dim> &fe_facet = scratch_data.fe_interface_values;
+      fe_facet.reinit(cell, f, sf, ncell, nf, nsf);
+      const auto &q_points = fe_facet.get_fe_values().get_quadrature_points();
```
Don't we also have `fe_facet.get_quadrature_points();` ? 
```suggestion
      const auto &q_points = fe_facet.get_quadrature_points();
```

tjhei commented:
I have more questions:
- In FEValues we differentiate the names in the main object and in the views: ``shape_value``  vs ``value``. Would it make sense to do the same here (shape_jump vs jump)?
- I just realized that we could rename ``choose`` as ``value`` (or shape_value). Looks more familiar, but maybe more confusing? I am not sure.

luca-heltai commented:
> I have more questions:

> In FEValues we differentiate the names in the main object and in the views: shape_value vs value. Would it make sense to do the same here (shape_jump vs jump)?

Yes, this would keep the interface consistent with what we have.

> I just realized that we could rename choose as value (or shape_value). Looks more familiar, but maybe more confusing? I am not sure.

I'm not sure about this. Choose is also somewhat arbitrary. Maybe value is better. What do others think?

tjhei commented:
> Yes, this would keep the interface consistent with what we have.

The disadvantage is, that a switch from scalar to vector PDE (with views) then requires switching again.

Thinking about this some more, I would use:
- shape_value(bool cell_or_neighbor, idx, q, component), shape_grad, ...
- jump(idx, q, component)
- average(idx, q, component)
in the views:
- value(cell_or_neighbor, idx, q)
- jump(idx, q), average(...), ...

But I will wait for somebody else to comment.

luca-heltai commented:
> Thinking about this some more, I would use:

>shape_value(bool cell_or_neighbor, idx, q, component), shape_grad, ...
jump(idx, q, component)
average(idx, q, component)
in the views:
value(cell_or_neighbor, idx, q)
jump(idx, q), average(...), ...

I like this proposal.

tjhei commented:
updated and ready to go from my end...

luca-heltai commented:
Ready to go also for me. Let's see what others think. If no-one objects, I'll merge tomorrow.

bangerth commented:
Let me take a look tomorrow before you merge!

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,690 @@
+ * objects. The class introduces a new "interface dof index" that walks over
+ * the union of the dof indices of the two FEFaceValues objects. Helper
+ * functions allow translating between the new "interface dof index" and the
+ * corresponding "fe index" and "dof index" of the FE.
```
can you explain what this "fe index" is? Does it relate to the active_fe_index in hp methods?

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,690 @@
+ * Discontinuous Galerkin methods on faces between two neighboring cells. In
+ * the literature these are called "inner interfaces" or "facets".
+ *
+ * Internally, this class provides an abstraction for two FEFaceValues
```
```suggestion
 * Internally, this class provides an abstraction for two FEFaceValues
 * or FESubfaceValues
```

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,690 @@
+  FEInterfaceValues(const FEFaceValues<dim> &   fe,
+                    const FESubfaceValues<dim> &fe_sub,
+                    const FEFaceValues<dim> &   fe_neighbor,
+                    const FESubfaceValues<dim> &fe_sub_neighbor);
```
Shouldn't there also be a constructor that just takes two FEValues objects? In fact, it's not clear to me why you need *four* objects here...

tjhei replied:
I think you are correct that there is no reason to supply different FE*Values for the neighbor here as the 4 internal objects are copied from the arguments anyways.

tjhei replied:
One could use this class to compute interface terms with different FEs on both sides as required for hp, but I have not tested this. Anyways, I changed the constructor.

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,689 @@
+      if (at_boundary())
+        return 0.0;
+      else
+        return -get_fe_values_neighbor().shape_value_component(shape_fct,
```
same question here: what happens for continuous functions? Do you get a negative here and a positive from the other side, that are then added up?

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,689 @@
+      if (at_boundary())
+        return 0.0;
+      else
+        return 0.5 * get_fe_values_neighbor().shape_value_component(shape_fct,
```
and here for continuous functions?

tjhei commented:
I leave some general comments down here instead of inline:
1. cell vs neighbor: Originally, I denoted the two cells adjacent to the interface as "cell" and "neighboring cell" as reflected in the members variables. I turned out to not be a useful naming convention as it can be confused with the two "neighboring cells of the interface". Instead, I am using "cell index" instead (and avoiding the "neighbor" term).
2. For continuous elements, you will have duplicate global dof indices in your "interface dof indices". While the definition of shape_value(), jump(), average() look discontinuous, the resulting integration will give continuous shape functions, zero jumps, and average() being equal to the single value. This class is mostly useful for DG elements(unless I implement derivatives, which will we added later).
2. sign of jump(): I think this shouldn't matter in practice, but my definition is consistent with Pietro/Ern 2012.

tjhei commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,689 @@
+  const FiniteElement<dim, spacedim> &fe,
+  const Quadrature<dim - 1> &         quadrature,
+  const UpdateFlags                   update_flags)
+  : internal_fe_face_values(StaticMappingQ1<dim, spacedim>::mapping,
```
Now I can no longer use it.

bangerth replied:
Why not? Can't just just forward to the other constructor that takes a mapping argument?

tjhei commented:
I think I addressed all your comments, thanks!

tjhei commented:
Well, after further experimentation with this, I think I need to make further changes:
For continuous elements, I am currently repeating dofindices in get_interface_dof_indices(). While correct, this is broken when used in AffineConstraints. The solution is deduplication of the indices, which means the interface has to change because there is no longer a one to one mapping of dofs and interface dofs. Will update soon.

bangerth commented:
I did think about this some and I think what you should be getting is a positive term and a negative term that in general will cancel each other once you add them up. (That's what you also described, I'm just writing this down for myself.)

Are you saying that once you hand a local matrix or vector along with a (non-unique) vector of DoF indices, that `AffineConstraints` messes up the result? You can of course fix this in two different ways: Make sure `AffineConstraints` can deal with duplicated DoF indices (likely a lot of work, given that class's extensive interface -- but also maybe worth fixing, or at least adding assertions to avoid wrong results); or de-duplicating DoFs in the new class. A final alternative would be to add a function `compress()` to `FEInterfaceValues` that takes a non-compressed DoF index vector, matrix, and vector, and compresses it by adding rows and columns of duplicated DoFs up and shrinking the object.

The most elegant solution would clearly be to de-duplicate before building matrices.

tjhei commented:
> did think about this some and I think what you should be getting is a positive term and a negative term that in general will cancel each other once you add them up.

Yes, for continuous elements jumps cancel out, averages are equal to the upstream and downstream values, but derivatives obviously do not.


> Are you saying that once you hand a local matrix or vector along with a (non-unique) vector of DoF indices, that `AffineConstraints` messes up the result?

Correct, and we don't have an assert that catches that. :-)

> The most elegant solution would clearly be to de-duplicate before building matrices.

Agreed. I have made an implementation of that (not pushed yet) and it works correctly, but it has one disadvantage: the number of interface dofs now can vary with every call to reinit(). For example, if we are at an refinement edge we might only share one dof. This is not intuitive, but I don't like any of the other options...

bangerth commented:
On 9/11/19 6:20 AM, Timo Heister wrote:
> Correct, and we don't have an assert that catches that. :-)

Ugh :-(


> Agreed. I have made an implementation of that (not pushed yet) and it 
> works correctly, but it has one disadvantage: the number of interface 
> dofs now can vary with every call to reinit(). For example, if we are at 
> an refinement edge we might only share one dof. This is not intuitive, 
> but I don't like any of the other options...

I think that's ok. It should still lead to a nice interface.

Let me know once you have something worth looking at again!

tjhei commented:
> Let me know once you have something worth looking at again!

I just pushed a first version that works well.

tjhei commented on tests/feinterface/fe_interface_values_07.output:
```diff
@@ -0,0 +1,11 @@
+DEAL::shape_value(true): 0.00000 0.500000 0.00000 0.500000 0.00000 0.00000 
+DEAL::shape_value(false): 0.00000 0.500000 0.00000 0.500000 0.00000 0.00000 
+DEAL::jump(): 0.00000 0.00000 0.00000 0.00000 0.00000 0.00000 
+DEAL::average(): 0.00000 0.500000 0.00000 0.500000 0.00000 0.00000 
```
You can see that for a Q1 element, we have jump=0, average=left=right.

tjhei commented:
@bangerth can you please have another look?

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * FEInterfaceValues is a data structure to access and assemble Finite Element
```
```suggestion
 * FEInterfaceValues is a data structure to access and assemble einite element
```

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+ * corresponding "cell index" (0 for the first cell, 1 for the second cell)
+ * and "dof index" within that cell.
+ *
+ * The class is made to be used inside MeshWorker::mesh_loop. It is intended
```
```suggestion
 * The class is made to be used inside MeshWorker::mesh_loop(). It is intended
```

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+   * on the lifetime of the arguments.
+   */
+  FEInterfaceValues(const FEFaceValues<dim> &   fe_face_values,
+                    const FESubfaceValues<dim> &fe_subface_values);
```
I think I still don't understand (i) what exactly is being copied, and (ii) whether the arguments have to be initialized to a particular face or subface.

bangerth replied:
Looking at the other constructor, I suspect that all you copy if the mapping, fe, quadrature, and update flags. But if that's what it is, I'd rather this constructor goes away altogether because I find it confusing that that is the first option. (Alternatively, move it below the other constructors and say that it forwards to these by copying out this that and the other.)

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+   * of two neighboring cells.
+   *
+   * The arguments (including their order) is identical to the @p face_worker arguments
+   * in MeshWorker::mesh_loop.
```
Just remove this sentence -- the function is easy enough to understand without having to look up something else. But if you want to keep it:
```suggestion
   * in MeshWorker::mesh_loop().
```
and replace is->are in the line above.

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+         const unsigned int &    sub_face_no_neighbor);
+
+  /**
+   * Re-initialize this object to be used on a interface given by a single face
```
```suggestion
   * Re-initialize this object to be used on an interface given by a single face
```

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+
+  /**
+   * Re-initialize this object to be used on a new interface given by two faces
+   * of two neighboring cells.
```
```suggestion
   * of two neighboring cells. The `cell` and `cell_neighbor` cells will be referred to through
   * `cell_index` zero and one after this call in all places where one needs to identify the
   * two cells adjacent to the interface.
```

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+   * Return a reference to the quadrature points in real space.
+   *
+   * @note This call requires a prior call to reinit() and having
+   * the @p update_quadrature_points flag set in the constructor.
```
We have a macro for this kind of thing: `@dealiiRequiresUpdateFlags{update_quadrature_points}` (see, for example, fe_values.cc, line 2928).

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+   * Return the number of DoFs (or shape functions) on the current interface.
+   *
+   * @note This number is only available after a call to reinit() and can change
+   * from one call to reinit() to the next.
```
Can you add one sentence about why it can change? Just a small example.

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+   * from one call to reinit() to the next.
+   */
+  unsigned
+  n_current_interface_dofs() const;
```
Move it down after `at_boundary()` so that it's close to similar functions.

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+   * function.
+   */
+  double
+  shape_value(const bool         use_upstream_value,
```
I think the function would be easier to understand if you named the argument something like `here_or_there`. The term "upstream" implies a direction, but no direction is provided -- you assume that the "stream direction" is the normal vector from cell 0, but then the (good) example in the last paragraph introduces a different direction vector.

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+              const unsigned int component = 0) const;
+
+  /**
+   * Return the jump $[u]=u_{\text{cell1}} - u_{\text{cell2}}$ on the
```
Maybe use "cell0" and "cell1", given that's how the `cell_index` is defined?

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+   * Return the jump $[u]=u_{\text{cell1}} - u_{\text{cell2}}$ on the
+   * interface
+   * for the shape function @p interface_dof_index at the quadrature point
+   * @p q_point of component @p component.
```
```suggestion
   * @p q_point of component @p component. Note that one can define the jump in
   * different ways (the value "there" minus the value "here", or the other way way
   * around; both are used in the finite element literature). The definition here uses
   * "value here minus value there", as seen from the first cell.
```

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+   * Pointer to internal_fe_face_values or internal_fe_subface_values,
+   * respectively as determined in reinit().
+   */
+  FEFaceValuesBase<dim> *fe_face_values = nullptr;
```
Just initialize it in the constructor as we usually do. Same for the next variable.

bangerth replied:
(actually, you already do that, so just drop the initializer here.)

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+
+  // Set up dof mapping and remove duplicates (for continuous elements).
+  // Possible optimization: if the FEs are discontinuous, we can just
+  // combine the vectors v and v2.
```
I don't see how you can reliably detect this. You might have a `{FE_RaviartThomas,FE_Nedelec,FE_Q,FE_DGQ}` combination -- some parts of it are continuous, others are not. I'd just not worry about it and remove the comment.

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+  for (unsigned int i = 0; i < interface_dof_indices.size(); ++i)
+    {
+      dofmap[i][0] = i;
+      dofmap[i][1] = numbers::invalid_unsigned_int;
```
Maybe
```
  dofmap[i] = {i, numbers::invalid_unsigned_int};
```
?

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+    return -get_fe_face_values(1).shape_value_component(dofpair[1],
+                                                        q_point,
+                                                        component);
+  return 0.0; // jump is zero for continuous shape functions
```
I don't think this is right. You are thinking that because the shape function appears to live on both cells, that it is continuous and so the . But that's not true. Think about the situation of an RT element: Its *normal* component is continuous, but its tangential component is not, even though it lives on both sides of the cell. I think you really have to compute the difference here.

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -0,0 +1,686 @@
+                                                             component);
+
+  // For continuous dof the average is equal to the value on the left:
+  return get_fe_face_values(0).shape_value_component(dofpair[0],
```
You have the same issue here.

tjhei commented:
> I think the jump and average functions are still not right, though.

You are right, thanks. This was certainly a bug. I think I addressed all your comments.

bangerth commented:
I was about to merge, but it's a sequence of 14 commits that each seem to patch each other up. Please squash as appropriate for a consistent history, and then merge yourself!
